dilation-protocol.md

```diff
@@ -388,158 +297,53 @@ In addition, the Leader must:
-delivered to the side that called `close()` before it sees
-`.connectionLost()`. Internally, the side which called `.loseConnection` will
-remain in a special "closing" state until the CLOSE response arrives, during
-which time DATA payloads are still delivered. After calling `close()` (or
-receiving CLOSE), any outbound `.write()` calls will trigger an error.
-
-DATA payloads that arrive for a non-open subchannel are logged and discarded.
-
-This protocol calls for one OPEN and two CLOSE messages for each subchannel,
-with some arbitrary number of DATA messages in between. Subchannel-ids should
-not be reused (it would probably work, the protocol hasn't been analyzed
-enough to be sure).
-
-The "control channel" is special. It uses a subchannel-id of all zeros, and
-is opened implicitly by both sides as soon as the first L3 connection is
-selected. It is routed to a special client-on-both-sides endpoint, rather
-than causing the listening endpoint to accept a new connection. This avoids
-the need for application-level code to negotiate who should be the one to
-open it (the Leader/Follower distinction is private to the Wormhole
-internals: applications are not obligated to pick a side).
-
-OPEN and CLOSE messages for the control channel are logged and discarded. The
-control-channel client endpoints can only be used once, and does not close
-until the Wormhole itself is closed.
-
-Each OPEN/DATA/CLOSE message is delivered to the L4 object for queueing,
-delivery, and eventual retirement. The L5 layer does not keep track of old
-messages.
+The DATA payloads are multiplexed over some virtual sub-channels. They may be
+exposed to the application as logically independent data streams.
+
+Each subchannel uses a distinct subchannel-id, which is a big-endian four-byte identifier.
+Both directions share a number space (unlike the sequence numbers),
+so the Dilation Leader uses odd numbers and the Follower uses even ones.
+Id `0` is special and used for the control channel.
+The control channel is always open; it cannot be opened nor closed explicitly.
+
+Sub-channels may be opened with a OPEN message and closed with CLOSE messages.
+Any side may close a channel. The other side, upon receiving that close message,
+should stop sending new messages and confirm with a CLOSE message too.
+
+DATA payloads that arrive for a non-open sub-channel should be ignored.
+TODO can we please remove that sentence? ~piegames
+Subchannel-ids should not be reused
+(it would probably work, the protocol hasn't been analyzed enough to be sure).
+
+The side that opens a sub-channel is named the Initiator, and the other side is named the Acceptor.
+Subchannels can be initiated in either direction, independent of the Leader/Follower distinction.
+For a typical file-transfer application, the subchannel would be initiated by the side seeking to send a file.
 
 ### Flow Control
 
+TODO review and potentially remove? ~piegames
```

Comment:
I would say: absolutely not, flow-control is pretty central to streaming applications.

Reply:
No, it would still say the stuff about pausing _ALL_ channels whenever a single ones needs to provide backpressure. That is, that implementations _should not_ attempt to provide their own subchannel flow-control, and instead rely on TCP's. (Competing flow-controllers on the same channel do not play nicely)

Reply:
Oh I see, I think I was confusing flow control and quality of service. (Tbf not entirely my fault, because the paragraph also talks about the one while actually describing the latter.)

So is the takeaway here "don't do your own flow control but feel free to implement QoS"?

Reply:
The original (https://github.com/magic-wormhole/magic-wormhole/blob/master/docs/dilation-protocol.md#flow-control) doesn't mention QoS at all, and I wouldn't mention it here.

Reply:
I know but it *talks* about QoS: "e.g. a chat conversation sharing a wormhole with file-transfer might prefer the IM text to take priority". Therefore why not talk about that explicitly?

Reply:
I guess that's certainly a use-case for QoS -- but what do you think is worth mentioning here? There's nothing to be done by implementers -- they have to pause "the TCP stream" to achieve backpressure.

If anything, this would be an interesting point for application developers: mixing high-bandwidth subchannels with low-bandwith ones may perform poorly under bad network conditions. But I think all they can do here is "not do that"?

That is, I don't believe that sentence is supposed to imply to Dilation implementers that they should attempt to implement some sort of QoS system. I don't recall details on how that'll interact (or not) with TCP windowing etc. (Usually QoS is done per-TCP-stream, e.g. prioritizing one whole stream over another).